renderers/react/src/render.tsx

```diff
@@ -32,16 +34,32 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
   return <Component {...args} />;
 };
 
+const WithCallback: FunctionComponent<{ callback: () => void; children: ReactElement }> = ({
+  callback,
+  children,
+}) => {
+  // See https://github.com/reactwg/react-18/discussions/5#discussioncomment-2276079
+  const once = useRef(false);
+  useLayoutEffect(() => {
+    if (once.current) return;
+    once.current = true;
+    callback();
+  }, [callback]);
+
+  return children;
+};
+
 const renderElement = async (node: ReactElement, el: Element) => {
   // Create Root Element conditionally for new React 18 Root Api
   const root = await getReactRoot(el);
 
   return new Promise((resolve) => {
     if (root) {
-      root.render(node);
-      setTimeout(() => {
-        resolve(null);
-      }, 0);
+      root.render(
+        <WithCallback key={Math.random()} callback={() => resolve(null)}>
```

Comment:
I had to add this `key={Math.random()}` to ensure the callback is called one time per `renderToDOM()` call. I'm not sure if that exactly cancels out the `useRef()` above and we could actually simplify the whole thing.

I'm not sure I quite understood @sebmarkbage's comment about strict mode properly, because in my experience refs are *not* shared across multiple renders in strict mode -- for example in this sandbox, the count is logged as `1` two times, that than "remembering": https://codesandbox.io/s/jovial-bartik-5llnzl?file=/src/App.js

I'd be inclined to simplify everything way down to just 

```js
useLayoutEffect(callback)
```

But I'd love to hear what others think @valentinpalkovic @chantastic (Chan is out this week so will need to wait until after that!)

Reply:
@tmeasday Could you tell me which comment by sebmarkbadge you mean? I cannot find the comment you are referring to.

My assumption is that a key property on the root element doesn‘t make much sense, because every time the React Dom render function is called a new component will be rendered. There shouldn‘t be any reference or connection to the previously rendered component.

Reply:
Arggh, sorry @valentinpalkovic I meant to link the comment (updated now), it's in the code also, but here it is:
https://github.com/reactwg/react-18/discussions/5#discussioncomment-2276079

So, the thought process on the PR as it currently stands is:

1. The `useRef` prevents the callback being triggered >1 times in strict mode for a single call to `renderToDOM`, following along the suggestion made in the comment.

2. However, if the component is retained between calls to `renderToDOM`, that means the callback is never triggered again so we recreate the component each time, to guarantee exactly one call to the callback for each call to `renderToDOM`.

What I'm not sure of is if simply doing `useLayoutEffect(callback, [])` will (in practice or guarantee) end up having the same effect.

Reply:
@tmeasday I think the example codesandbox isn't appropriate.

A better example to test strict mode and multiple executions of useEffect/useLayoutEffect is this one: https://codesandbox.io/s/autumn-platform-0rpc3f?file=/src/App.js

Although 0 is returned, the console.log clearly shows the iteration of the useRef value. A rerendering of course doesn't take place, because we are mutating ref.current, which doesn't trigger a rerendering.

Reply:
@tmeasday 

`However, if the WithCallback component is retained between calls to renderToDOM` -> I couldn't find any information about this. Can you point me in the right direction? Is this just an assumption, or did you read that somewhere?

Reply:
@valentinpalkovic what I mean is if you do:

```js
root.render(<Component />)

// Then later
root.render(<Component />)
```

Then React does not reinstantiate `Component`, assuming that the variable here has not changed and is referentially equal (`===`). In particular, `useEffect(..., [])` will not run again. If you add a random `key` it will of course.

You can see it in action here: https://codesandbox.io/s/young-waterfall-zk8qeb?file=/src/index.js

Reply:
As for where it is documented, I guess here? https://reactjs.org/docs/reconciliation.html#component-elements-of-the-same-type

Reply:
One thing I should check is that the children of `WithCallback` are not also re-instantiated, which would break a bunch of things 🤔

Reply:
Understood! The question is whether `unmountElement` is called before `renderElement` is called once again with another Story. The `key` property shouldn't be necessary if that's the case. It is only necessary if `root.render` is called on the same instance. But if a new instance of `root` is provided, every time the `renderElement` function is called, the `key` property shouldn't be necessary. If this assumption is not the case, we should instead figure out why the timing of `unmountElement` isn't appropriate.

Reply:
> If this assumption is not the case, we should instead figure out why the timing of unmountElement isn't appropriate.

Ahh, so there are scenarios where you want to call unmount and other scenarios where you do not. For instance when changing `args` we don't want to remount the story/component, whereas when you change story, we do (that's what users expect in any case it seems, you could argue differently).

That's what we use the `forceRemount` flag to control.